Python/Product/VSInterpreters/PackageManager/CondaUtils.cs

```diff
@@ -68,43 +72,104 @@ static class CondaUtils {
+        /// <remarks>Result is cached, it is safe to call multiple times.</remarks>
+        internal static Task<IEnumerable<KeyValuePair<string, string>>> GetActivationEnvironmentVariablesForPrefixAsync(string condaPath, string prefixPath) {
+            return CaptureActivationEnvironmentVariablesAsync(condaPath, prefixPath);
         }
 
         /// <summary>
         /// Activate the specified conda environment, capture and return its environment variables.
         /// </summary>
-        /// <param name="condaPath">Path to the base conda environment's conda.exe</param>
+        /// <param name="condaPath">Path to the root conda environment's conda.exe</param>
         /// <param name="prefixPath">Path to the conda environment to activate.</param>
         /// <returns>List of environment variables.</returns>
-        internal static IEnumerable<KeyValuePair<string, string>> CaptureActivationEnvironmentVariablesForPrefix(string condaPath, string prefixPath) {
-            var activateBat = Path.Combine(Path.GetDirectoryName(condaPath), "activate.bat");
-            if (File.Exists(activateBat)) {
-                using (var proc = ProcessOutput.RunHiddenAndCapture(activateBat, new[] { prefixPath, "&", "python.exe", "-c", PrintEnvironmentCode })) {
-                    proc.Wait();
-                    if (proc.ExitCode == 0) {
-                        return ParseEnvironmentVariables(proc);
+        /// <remarks>Result is cached, it is safe to call multiple times.</remarks>
+        internal static IEnumerable<KeyValuePair<string, string>> GetActivationEnvironmentVariablesForPrefix(string condaPath, string prefixPath) {
+            return CaptureActivationEnvironmentVariables(condaPath, prefixPath);
+        }
+
+        private async static Task<IEnumerable<KeyValuePair<string, string>>> CaptureActivationEnvironmentVariablesAsync(string condaPath, string prefixPath) {
+            using (await _activationCacheLock.LockAsync(CancellationToken.None)) {
+                var condaKey = new CondaCacheKey(condaPath, prefixPath);
+
+                if (!_activationCache.TryGetValue(condaKey, out KeyValuePair<string, string>[] activationVariables)) {
+                    activationVariables = null;
+
+                    using (var proc = StartCaptureProcess(condaPath, prefixPath)) {
+                        if (proc != null) {
+                            await proc;
+                            if (proc.ExitCode == 0) {
+                                activationVariables = ParseEnvironmentVariables(proc).ToArray();
+                            }
+                        }
+                    }
+
+                    _activationCache[condaKey] = activationVariables ?? new KeyValuePair<string, string>[0];
+                }
+
+                return activationVariables;
+            }
+        }
+
+        private static IEnumerable<KeyValuePair<string, string>> CaptureActivationEnvironmentVariables(string condaPath, string prefixPath) {
```

Comment:
do we need the non-async version?

can't we just have the sync code use the async version with .GetAwaiter().GetResult()   ?

Reply:
`.GetAwaiter().GetResult()` is generally not recommended but in both cases where I need the sync function, I have access to `IServiceProvider`, which means I could use `site.GetUIThread().InvokeTaskSync` to invoke the async method and block safely. I'll make the change.